### PR TITLE
Datastream stop transition redesign

### PR DESCRIPTION
**Pull request by shrinandthakkar**
Introducing a new intermediate STOPPING state in the lifecycle of a datastream, which represents a transitioning state for the leader to calculate and generate a new reassignment when a stop request is issued. This blocking transitioning state would make sure that any updates made to the datastream are applied successfully if datastream is restarted with a stop and resume call with a 60-second timeout in between. 

Important: DO NOT REPORT SECURITY ISSUES DIRECTLY ON GITHUB.  
For reporting security issues and contributing security fixes,  
please, email security@linkedin.com instead, as described in  
the contribution guidelines.

Please, take a minute to review the contribution guidelines at:  
https://github.com/linkedin/Brooklin/blob/master/CONTRIBUTING.md

### CODE REVIEW

File: datastream-server/src/main/java/com/linkedin/datastream/server/Coordinator.java [hunk 1]
```diff
@@ -1084,14 +1087,16 @@ private void deleteTopic(Datastream datastream) {
...
     List<String> liveInstances = Collections.emptyList();
     Map<String, Set<DatastreamTask>> previousAssignmentByInstance = Collections.emptyMap();
     Map<String, List<DatastreamTask>> newAssignmentsByInstance = Collections.emptyMap();
 
     try {
       List<DatastreamGroup> datastreamGroups = fetchDatastreamGroups();
-
+      List<DatastreamGroup> stoppingDatastreamGroups =
+          fetchDatastreamGroupsWithStatus(Collections.singletonList(DatastreamStatus.STOPPING));
+      _log.debug("stopping datastreams: {}", stoppingDatastreamGroups);
```
Comment (vmaheshw) [inline datastream-server/src/main/java/com/linkedin/datastream/server/Coordinator.java:1099]: This should be an `INFO` level logging. This is an important log message to debug the datastream state change. Also, `DatastreamGroup.toString()` method is an optimized method that prints only the datastream names, so it will be a small log message. You can check the size of the list and print only when it is non-zero.

File: datastream-server/src/main/java/com/linkedin/datastream/server/Coordinator.java [hunk 2]
```diff
@@ -1114,16 +1119,24 @@ private void handleLeaderDoAssignment(boolean cleanUpOrphanNodes) {
       // it failed to create or delete zNodes), we will do our best to continue the current process
       // and schedule a retry. The retry should be able to diff the remaining ZooKeeper work
       _adapter.updateAllAssignments(newAssignmentsByInstance);
+
+      stoppingDatastreamGroups.forEach(datastreamGroup -> datastreamGroup.getDatastreams().forEach(datastream -> {
```
Comment (vmaheshw) [inline datastream-server/src/main/java/com/linkedin/datastream/server/Coordinator.java:1123]: For readability purpose, it is a good practice to have 1 for loop per line. I understand that `forEach` has its benefits, but in this case, a regular for loop will avoid the use of `AtomicInteger` which is an over-kill.

File: datastream-server/src/main/java/com/linkedin/datastream/server/Coordinator.java [hunk 3]
```diff
@@ -1084,14 +1087,16 @@ private void deleteTopic(Datastream datastream) {
    * any instance after old unused tasks are cleaned up.
    */
   private void handleLeaderDoAssignment(boolean cleanUpOrphanNodes) {
-    boolean succeeded = true;
+    AtomicBoolean succeeded = new AtomicBoolean(true);
```
Comment (vmaheshw) [inline datastream-server/src/main/java/com/linkedin/datastream/server/Coordinator.java:1090]: See if you can avoid moving to AtomicBoolean.

File: datastream-server-restli/src/main/java/com/linkedin/datastream/server/dms/DatastreamResources.java [hunk 4]
```diff
@@ -406,28 +410,40 @@ public UpdateResponse update(String key, Datastream datastream) {
...
+            "Could not update datastream to STOPPING state: " + d.getName(), e);
       }
     }
 
-    LOG.info("Completed request for stopping datastream {}", datastream);
+    // polls until the leader transitions the state of the datastream to STOPPED state
+    PollUtils.poll(() -> datastreamsToStop.stream()
+            .allMatch(ds -> _store.getDatastream(ds.getName()).getStatus().equals(DatastreamStatus.STOPPED)),
+        allStopped -> allStopped, TRANSITION_RETRY_PERIOD_MS_DEFAULT.toMillis(),
+        TRANSITION_TIMEOUT_MS_DEFAULT.toMillis()).orElseThrow(RetriesExhaustedException::new);
```
Comment (vmaheshw) [inline datastream-server-restli/src/main/java/com/linkedin/datastream/server/dms/DatastreamResources.java:444]: What will be the Status code returned after the timeout? Can you define a deterministic behavior and return appropriate error message and error code?

File: datastream-server-restli/src/test/java/com/linkedin/datastream/server/dms/TestDatastreamResources.java [hunk 5]
```diff
@@ -230,6 +237,147 @@ public void testPauseDatastream() {
...
+    Assert.assertEquals(stopResponse.getStatus(), HttpStatus.S_200_OK);
+
+    // Retrieve datastream and check that is in STOPPED state.
+    Datastream ds = resource2.get(datastreamName);
+    Assert.assertNotNull(ds);
+    Assert.assertEquals(ds.getStatus(), DatastreamStatus.STOPPED);
+  }
+
+  @Test
+  public void testStopDatastreamSimultaneous() throws InterruptedException {
```
Comment (vmaheshw) [inline datastream-server-restli/src/test/java/com/linkedin/datastream/server/dms/TestDatastreamResources.java:273]: please fix the name. The description is not clear.

Comment (vmaheshw) [reply datastream-server-restli/src/test/java/com/linkedin/datastream/server/dms/TestDatastreamResources.java:273]: you can remove `InterruptedException`.

File: datastream-server-restli/src/test/java/com/linkedin/datastream/server/dms/TestDatastreamResources.java [hunk 6]
```diff
@@ -230,6 +237,147 @@ public void testPauseDatastream() {
...
+    }
+
+    // Retrieve datastream and check that is in STOPPED state.
+    Datastream ds = resource2.get(datastreamName);
+    Assert.assertNotNull(ds);
+    Assert.assertEquals(ds.getStatus(), DatastreamStatus.STOPPED);
+  }
+
+  @Test
+  public void testStopDatastreamRequestHandlerHostDies() {
```
Comment (vmaheshw) [inline datastream-server-restli/src/test/java/com/linkedin/datastream/server/dms/TestDatastreamResources.java:314]: Plz fix the name of the test. probably `HostDies` to `WhenTheHostDies`

File: datastream-server-restli/src/test/java/com/linkedin/datastream/server/dms/TestDatastreamResources.java [hunk 7]
```diff
@@ -230,6 +237,147 @@ public void testPauseDatastream() {
...
+    // If the datastream is found in STOPPED state, assuming that the leader already transitioned the datastream from STOPPING to STOPPED.
+    if (retrievedStatus.equals(DatastreamStatus.STOPPED)) {
+      return;
+    }
+    Assert.assertEquals(retrievedStatus, DatastreamStatus.STOPPING);
+
+    // If the datastream is found in STOPPING state, there could be two cases here,
+    // 1. The leader is notified before the host crashed, in that case the leader would transition the datastream back to STOPPED state.
+    //    Hence polling in try block until the leader transitions the state of the datastream to STOPPED state.
+
```
Comment (vmaheshw) [inline datastream-server-restli/src/test/java/com/linkedin/datastream/server/dms/TestDatastreamResources.java:362]: nit: empty line

File: datastream-server-restli/src/test/java/com/linkedin/datastream/server/dms/TestDatastreamResources.java [hunk 8]
```diff
@@ -47,14 +53,15 @@
 import com.linkedin.restli.server.PathKeys;
 import com.linkedin.restli.server.RestLiServiceException;
 
-
 /**
  * Test DatastreamResources with ZooKeeper backed DatastreamStore
  */
 @Test(singleThreaded = true)
 public class TestDatastreamResources {
```
Comment (vmaheshw) [inline datastream-server-restli/src/test/java/com/linkedin/datastream/server/dms/TestDatastreamResources.java:60]: Some of the unit-tests are missing:

a) Trying to stop the datastream in STOPPING state.
b) Trying to pause/resume the datastream in STOPPING state.
c) Trying to delete the datastream in STOPPING state.
d) Leader busy and not able to pick the LEADER_DO_ASSIGNMENT message or busy updating the zookeeper, and the request timed-out.

**Review by vmaheshw — CHANGES_REQUESTED** [review]

File: datastream-server-restli/src/test/java/com/linkedin/datastream/server/dms/TestDatastreamResources.java (hunk 8 above)
Comment (shrinandthakkar) [reply datastream-server-restli/src/test/java/com/linkedin/datastream/server/dms/TestDatastreamResources.java:61]: I have added tests a, b & c. Any suggestions on how can we test the (d) scenario in unit tests?

File: datastream-server-restli/src/main/java/com/linkedin/datastream/server/dms/DatastreamResources.java [hunk 9]
```diff
@@ -96,6 +98,8 @@
   private static final Supplier<Long> DELETE_CALL_LATENCY_MS_SUPPLIER = DELETE_CALL_LATENCY_MS::get;
   private static final String CREATE_CALL_LATENCY_MS_STRING = "createCallLatencyMs";
   private static final String DELETE_CALL_LATENCY_MS_STRING = "deleteCallLatencyMs";
+  private static final Duration TRANSITION_TIMEOUT_MS_DEFAULT = Duration.ofMillis(1000);
```
[the remark this answers is not in the document]
Comment (shrinandthakkar) [reply datastream-server-restli/src/main/java/com/linkedin/datastream/server/dms/DatastreamResources.java:100]: I don't think if there's any issue with a longer timeout. What should be the practical timeout period?

File: datastream-server-restli/src/test/java/com/linkedin/datastream/server/dms/TestDatastreamResources.java (hunk 8 above)
Comment (vmaheshw) [reply datastream-server-restli/src/test/java/com/linkedin/datastream/server/dms/TestDatastreamResources.java:61]: for d) You can mock the leader coordinator, mark `fetchDatastreamGroupsWithStatus` `visibleForTesting` and return empty list when this api is called with STOPPING status.

Another thing you can do is extract a method out of fetchDatastreamGroupsWithStatus(Collections.singletonList(DatastreamStatus.STOPPING)); and return empty list.

File: datastream-server-restli/src/test/java/com/linkedin/datastream/server/dms/TestDatastreamResources.java [hunk 10]
```diff
@@ -547,8 +745,9 @@ public void testUpdateDatastream() throws Exception {
     Assert.assertTrue(PollUtils.poll(() -> {
       Datastream updatedDatastream1 = resource.get(datastream1.getName());
       Datastream updatedDatastream2 = resource.get(datastream2.getName());
-      return updatedDatastream1.getMetadata().get("key").equals("value1")
-          && updatedDatastream2.getMetadata().get("key").equals("value2");
+      return updatedDatastream1.getMetadata().get("key").equals("value1") && updatedDatastream2.getMetadata()
+          .get("key")
+          .equals("value2");
```
Comment (vmaheshw) [inline datastream-server-restli/src/test/java/com/linkedin/datastream/server/dms/TestDatastreamResources.java:750]: why is this change required ? Plz revert it for indentation reasons.

File: datastream-server-restli/src/test/java/com/linkedin/datastream/server/dms/TestDatastreamResources.java (hunk 8 above)
Comment (shrinandthakkar) [reply datastream-server-restli/src/test/java/com/linkedin/datastream/server/dms/TestDatastreamResources.java:62]: It was harder to mock the coordinator, so mocked the state store to achieve this scenario.

File: datastream-server-restli/src/main/java/com/linkedin/datastream/server/dms/DatastreamResources.java [hunk 11]
```diff
@@ -97,33 +99,57 @@
...
+
   private final DatastreamStore _store;
   private final Coordinator _coordinator;
   private final ErrorLogger _errorLogger;
 
   private final DynamicMetricsManager _dynamicMetricsManager;
 
+  private Duration _stopTransitionTimeoutMs = STOP_TRANSITION_TIMEOUT_MS_DEFAULT;
+  private Duration _stopTransitionRetryPeriodMs = STOP_TRANSITION_RETRY_PERIOD_MS_DEFAULT;
+
```
Comment (vmaheshw) [inline datastream-server-restli/src/main/java/com/linkedin/datastream/server/dms/DatastreamResources.java:116]: Can you convert these variables to `final` to avoid their modification in any other place other than the constructor

File: datastream-server-restli/src/main/java/com/linkedin/datastream/server/dms/DatastreamResources.java [hunk 12]
```diff
@@ -97,33 +99,57 @@
...
     _dynamicMetricsManager.registerGauge(CLASS_NAME, DELETE_CALL_LATENCY_MS_STRING, DELETE_CALL_LATENCY_MS_SUPPLIER);
+
+    if (Objects.isNull(properties)) {
+      return;
+    }
+    if (properties.containsKey(CONFIG_STOP_TRANSITION_TIMEOUT_MS)) {
+      _stopTransitionTimeoutMs = Duration.ofMillis(Long.parseLong(properties.getProperty(CONFIG_STOP_TRANSITION_TIMEOUT_MS)));
+    }
+    if (properties.containsKey(CONFIG_STOP_TRANSITION_RETRY_PERIOD_MS)) {
+      _stopTransitionRetryPeriodMs = Duration.ofMillis(Long.parseLong(properties.getProperty(CONFIG_STOP_TRANSITION_RETRY_PERIOD_MS)));
```
Comment (vmaheshw) [inline datastream-server-restli/src/main/java/com/linkedin/datastream/server/dms/DatastreamResources.java:146]: nit: you can use `getOrDefault` method to summarize these if statements.

File: datastream-server-restli/src/main/java/com/linkedin/datastream/server/dms/DatastreamResources.java [hunk 13]
```diff
@@ -97,33 +99,57 @@
...
+    }
+    Validate.isTrue(_stopTransitionRetryPeriodMs.toMillis() > 0);
+    Validate.isTrue(_stopTransitionTimeoutMs.toMillis() > _stopTransitionRetryPeriodMs.toMillis());
```
Comment (vmaheshw) [inline datastream-server-restli/src/main/java/com/linkedin/datastream/server/dms/DatastreamResources.java:149]: Just a thought: A bad config can literally bring down the cluster in this case. What do you think in terms of the cluster operability ? One thing I can think of logging the warning and setting the defaults to avoid the bad config bringing down the cluster.

File: datastream-server-restli/src/main/java/com/linkedin/datastream/server/dms/DatastreamResources.java [hunk 14]
```diff
@@ -404,30 +430,50 @@ public UpdateResponse update(String key, Datastream datastream) {
...
+            "Could not update datastream to STOPPING state: " + d.getName(), e);
       }
     }
 
-    LOG.info("Completed request for stopping datastream {}", datastream);
+    // polls until the leader transitions the state of the datastream to STOPPED state
+    PollUtils.poll(() -> datastreamsToStop.stream()
+            .allMatch(ds -> _store.getDatastream(ds.getName()).getStatus().equals(DatastreamStatus.STOPPED)),
+        allStopped -> allStopped, _stopTransitionRetryPeriodMs.toMillis(),
+        _stopTransitionTimeoutMs.toMillis())
```
Comment (vmaheshw) [inline datastream-server-restli/src/main/java/com/linkedin/datastream/server/dms/DatastreamResources.java:472]: nit pick: you can merge these two lines. (Feel free to ignore it)

File: datastream-server-restli/src/main/java/com/linkedin/datastream/server/dms/DatastreamResources.java [hunk 15]
```diff
@@ -97,33 +99,57 @@
...
+    this(datastreamServer.getDatastreamStore(), datastreamServer.getCoordinator(), datastreamServer.getProperties());
   }
 
   /**
    * Constructor for DatastreamResources
    * @param store the datastream store
    * @param coordinator the server coordinator
    */
-  public DatastreamResources(DatastreamStore store, Coordinator coordinator) {
+  public DatastreamResources(DatastreamStore store, Coordinator coordinator, Properties properties) {
```
Comment (vmaheshw) [inline datastream-server-restli/src/main/java/com/linkedin/datastream/server/dms/DatastreamResources.java:130]: nit: You could have created another construction marked as @VisibleForTesting with store and coordinator fields and call DatastreamResources(store, coordinator, null); This would have avoided most of the test changes required for this new definition.

**Review by vmaheshw — APPROVED** [review]
Mostly LGTM except some nit picks.

File: datastream-server-restli/src/main/java/com/linkedin/datastream/server/dms/DatastreamResources.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (shrinandthakkar) [reply datastream-server-restli/src/main/java/com/linkedin/datastream/server/dms/DatastreamResources.java]: Increased the default timeout and made it configurable which could be set particular to the cluster.

**Review by vmaheshw — APPROVED** [review]

**Review by jzakaryan — APPROVED** [review]
lgtm
